File: js/src/cloud_bucket_mount.ts
import {
  CloudBucketMount_BucketType,
  CloudBucketMount as CloudBucketMountProto,
} from "../proto/modal_proto/api";
import { ModalClient } from "./client";
import { Secret } from "./secret";

export class CloudBucketMountService {
  readonly #client: ModalClient;

  constructor(client: ModalClient) {
    this.#client = client;
  }

  create(
    bucketName: string,
    params: {
      secret?: Secret;
      readOnly?: boolean;
      requesterPays?: boolean;
      bucketEndpointUrl?: string;
      keyPrefix?: string;
      oidcAuthRoleArn?: string;
    } = {},
  ): CloudBucketMount {
    let bucketType = CloudBucketMount_BucketType.S3;
    if (params.bucketEndpointUrl) {
      const url = new URL(params.bucketEndpointUrl);
      if (url.hostname.endsWith("r2.cloudflarestorage.com")) {
        bucketType = CloudBucketMount_BucketType.R2;
      } else if (url.hostname.endsWith("storage.googleapis.com")) {
        bucketType = CloudBucketMount_BucketType.GCP;
      } else {
        bucketType = CloudBucketMount_BucketType.S3;
        this.#client.logger.debug(
          "CloudBucketMount received unrecognized bucket endpoint URL. " +
            "Assuming AWS S3 configuration as fallback.",
          "bucketEndpointUrl",
          params.bucketEndpointUrl,
        );
      }
    }

    if (params.requesterPays && !params.secret) {
      throw new Error("Credentials required in order to use Requester Pays.");
    }

    if (params.keyPrefix && !params.keyPrefix.endsWith("/")) {
      throw new Error(
        "keyPrefix will be prefixed to all object paths, so it must end in a '/'",
      );
    }

    return new CloudBucketMount(
      bucketName,
      params.secret,
      params.readOnly ?? false,
      params.requesterPays ?? false,
      params.bucketEndpointUrl,
      params.keyPrefix,
      params.oidcAuthRoleArn,
      bucketType,
    );
  }
}

/** Cloud Bucket Mounts provide access to cloud storage buckets within Modal Functions. */
export class CloudBucketMount {
  readonly bucketName!: string;
  readonly secret?: Secret;
  readonly readOnly!: boolean;
  readonly requesterPays!: boolean;
  readonly bucketEndpointUrl?: string;
  readonly keyPrefix?: string;
  readonly oidcAuthRoleArn?: string;
  readonly #bucketType!: CloudBucketMount_BucketType;

  /** @ignore */
  constructor(
    bucketName: string,
    secret: Secret | undefined,
    readOnly: boolean,
    requesterPays: boolean,
    bucketEndpointUrl: string | undefined,
    keyPrefix: string | undefined,
    oidcAuthRoleArn: string | undefined,
    bucketType: CloudBucketMount_BucketType,
  ) {
    this.bucketName = bucketName;
    this.secret = secret;
    this.readOnly = readOnly;
    this.requesterPays = requesterPays;
    this.bucketEndpointUrl = bucketEndpointUrl;
    this.keyPrefix = keyPrefix;
    this.oidcAuthRoleArn = oidcAuthRoleArn;
    this.#bucketType = bucketType;
  }

  /** @ignore */
  toProto(mountPath: string): CloudBucketMountProto {
    return CloudBucketMountProto.create({
      bucketName: this.bucketName,
      mountPath,
      credentialsSecretId: this.secret?.secretId ?? "",
      readOnly: this.readOnly,
      bucketType: this.#bucketType,
      requesterPays: this.requesterPays,
      bucketEndpointUrl: this.bucketEndpointUrl,
      keyPrefix: this.keyPrefix,
      oidcAuthRoleArn: this.oidcAuthRoleArn,
    });
  }
}

File: go/cloud_bucket_mount.go
package modal

import (
	"context"
	"fmt"
	"net/url"
	"strings"

	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
)

// CloudBucketMountService provides CloudBucketMount related operations.
type CloudBucketMountService interface {
	New(bucketName string, params *CloudBucketMountParams) (*CloudBucketMount, error)
}

type cloudBucketMountServiceImpl struct{ client *Client }

// CloudBucketMount provides access to cloud storage buckets within Modal Functions.
type CloudBucketMount struct {
	BucketName        string
	Secret            *Secret
	ReadOnly          bool
	RequesterPays     bool
	BucketEndpointURL *string
	KeyPrefix         *string
	OidcAuthRoleArn   *string
	bucketType        pb.CloudBucketMount_BucketType
}

// CloudBucketMountParams are options for creating a CloudBucketMount.
type CloudBucketMountParams struct {
	Secret            *Secret
	ReadOnly          bool
	RequesterPays     bool
	BucketEndpointURL *string
	KeyPrefix         *string
	OidcAuthRoleArn   *string
}

// New creates a new CloudBucketMount.
func (s *cloudBucketMountServiceImpl) New(bucketName string, params *CloudBucketMountParams) (*CloudBucketMount, error) {
	ctx := context.Background()
	if params == nil {
		params = &CloudBucketMountParams{}
	}

	mount := &CloudBucketMount{
		BucketName:        bucketName,
		Secret:            params.Secret,
		ReadOnly:          params.ReadOnly,
		RequesterPays:     params.RequesterPays,
		BucketEndpointURL: params.BucketEndpointURL,
		KeyPrefix:         params.KeyPrefix,
		OidcAuthRoleArn:   params.OidcAuthRoleArn,
	}

	if mount.BucketEndpointURL != nil {
		parsedURL, err := url.Parse(*mount.BucketEndpointURL)
		if err != nil {
			return nil, fmt.Errorf("invalid bucket endpoint URL: %w", err)
		}

		hostname := parsedURL.Hostname()
		if strings.HasSuffix(hostname, "r2.cloudflarestorage.com") {
			mount.bucketType = pb.CloudBucketMount_R2
		} else if strings.HasSuffix(hostname, "storage.googleapis.com") {
			mount.bucketType = pb.CloudBucketMount_GCP
		} else {
			mount.bucketType = pb.CloudBucketMount_S3
			if s.client != nil && s.client.logger != nil {
				s.client.logger.DebugContext(
					ctx,
					"CloudBucketMount received unrecognized bucket endpoint URL. Assuming AWS S3 configuration as fallback.",
					"BucketEndpointURL", *mount.BucketEndpointURL,
				)
			}
		}
	} else {
		mount.bucketType = pb.CloudBucketMount_S3
	}

	if mount.RequesterPays && mount.Secret == nil {
		return nil, fmt.Errorf("credentials required in order to use Requester Pays")
	}

	if mount.KeyPrefix != nil && !strings.HasSuffix(*mount.KeyPrefix, "/") {
		return nil, fmt.Errorf("keyPrefix will be prefixed to all object paths, so it must end in a '/'")
	}

	return mount, nil
}

func (c *CloudBucketMount) toProto(mountPath string) (*pb.CloudBucketMount, error) {
	credentialsSecretID := ""
	if c.Secret != nil {
		credentialsSecretID = c.Secret.SecretID
	}

	return pb.CloudBucketMount_builder{
		BucketName:          c.BucketName,
		MountPath:           mountPath,
		CredentialsSecretId: credentialsSecretID,
		ReadOnly:            c.ReadOnly,
		BucketType:          c.bucketType,
		RequesterPays:       c.RequesterPays,
		BucketEndpointUrl:   c.BucketEndpointURL,
		KeyPrefix:           c.KeyPrefix,
		OidcAuthRoleArn:     c.OidcAuthRoleArn,
	}.Build(), nil
}
